### PR TITLE
Implemented task

### PR DESCRIPTION
**Pull request by OlegMR0**

### CODE REVIEW

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+import mate.academy.rickandmorty.dto.internal.CharacterDto;
+import mate.academy.rickandmorty.model.Character;
+import mate.academy.rickandmorty.service.CharacterService;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Rick and Morty API")
+@RestController
+@RequestMapping("/api")
```
Comment (olekskov) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]: i'd expand path in order to be able to add other endpoints later
```suggestion
@RequestMapping("/api/characters")

```

File: src/main/java/mate/academy/rickandmorty/repository/CharacterRepository.java [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface CharacterRepository extends JpaRepository<Character, Long> {
+    long count();
+
+    Character save(Character character);
+
+    List<Character> findAll();
+
+    Optional<Character> findById(Long id);
```
Comment (olekskov) [inline src/main/java/mate/academy/rickandmorty/repository/CharacterRepository.java]: these methods are present by default in JpaRepository, no need to duplicate
```suggestion
```

File: src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+public class CharacterServiceImpl implements CharacterService {
+    private CharacterClient client;
+    private CharacterRepository repository;
+    private CharacterMapper mapper;
+
+    @PostConstruct
+    private void init() {
+        List<Character> characters = client.getCharacters();
+        repository.saveAll(characters);
+    }
```
Comment (olekskov) [inline src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java]: Please move method to CharacterClientImpl
This service mainly works to process our users' requests and initialization is logically out of this scope
```suggestion
```

File: src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public Character saveRandomCharacter() {
+        long bound = repository.count();
+        long randomLong = new Random().nextLong(bound);
+        Character character = repository.findById(randomLong)
+                .orElseThrow(() -> new RuntimeException(
+                        "Occurred an error while pulling random character"));
+        character.setInternalId(null);
+        return repository.save(character);
```
Comment (olekskov) [inline src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java]: Cannot understand why we reset id and call save for character we just retrieved from DB

File: src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+    @Override
+    public List<CharacterDto> getCharactersByName(String name) {
+        List<Character> characters = repository.findByNameContaining(name);
+        List<CharacterDto> characterDtos = mapper.map(characters);
+        return characterDtos;
+    }
+
+    @Override
+    public Character saveRandomCharacter() {
```
Comment (olekskov) [inline src/main/java/mate/academy/rickandmorty/service/CharacterServiceImpl.java]: you should unify if you service returns DTOs or models

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    @Operation(summary = "Find characters by specific name.")
+    @GetMapping("/search")
+    public List<CharacterDto> getCharactersByName(String name) {
+        return characterService.getCharactersByName(name);
+    }
+
+    @Operation(summary = "Generate and save random existing character.")
+    @GetMapping("/random")
+    public Character saveRandomCharacter() {
```
Comment (olekskov) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java]: please use dtos and hide internal implementation of classes

**Review by olekskov — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/rickandmorty/controller/CharacterController.java [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+@RestController
+@RequestMapping("/api/characters")
+@AllArgsConstructor
+public class CharacterController {
+    private CharacterService characterService;
+
+    @Operation(summary = "Get all available characters.")
+    @GetMapping
+    public List<CharacterDto> getCharacters() {
+        return characterService.getAll();
```
Comment (Rommelua) [inline src/main/java/mate/academy/rickandmorty/controller/CharacterController.java:23]: Have you ever heard about YAGNI?
image

**Review by Rommelua — APPROVED** [review]

**Pull request closed without merging**